lib/patientFinder/utils.rb

```diff
@@ -0,0 +1,142 @@
+        
+        #find matches for test patients
+        def self.get_MRNs(master_records, test_patients)
+    
+            #store the results
+            tests_to_MRNs= Hash.new
+            
+            #iterate through master records and find candidates 
+            test_patients.each { |test_patient|
+                #each test patient might have one or more MRN candidates
+                candidates = get_candidates_for_test(master_records, test_patient.record)
+                
+                #resolve cadidates to unique MRNs - with other words resolve the candidates
+                tests_to_MRNs[test_patient.fileName]=get_MRN_for_tests(candidates.map(&:medical_record_number))
+            }
+    
+            tests_to_MRNs
+        end
+        
+        #resolves ambiguos candidates to a unique MRN 
+        def self.get_MRN_for_ambiguous(candidate_MRNs)
+            AMBIGUOUS_PATIENTS.each_pair {  |ambiguous_entries, mrn| 
+                return mrn if (candidate_MRNs-ambiguous_entries).empty?
+            }
+            "too ambiguous #{candidate_MRNs}"
+        end
+    
+        
+        #resolves the candidates - if one single candidate found then retunr that - if more then resolve them through disambiguation
+        def self.get_MRN_for_tests(candidate_MRNs)
+            
+            return candidate_MRNs.first if(candidate_MRNs.length==1)
+            return get_MRN_for_ambiguous(candidate_MRNs) if (candidate_MRNs.length>1)
+        end
+        
+        #a utility that iterates through a set of master records to find possible matches (candidates) for a test cord
+        def self.get_candidates_for_test(master_records,test_record)
+            
+            candidates=Array.new
+            # candidates
+            master_records.each { |master_record|
+            
+                #simple conditions that will reduce the record pool
+                if(master_record.gender==test_record.gender &&
+                    master_record.ethnicity[:code]==test_record.ethnicity[:code] &&
+                    master_record.race[:code] == test_record.race[:code])
+                    
+                    #go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+                    sections = [:allergies, :care_goals, :conditions, :encounters, :immunizations, :medical_equipment, :medications, :procedures, :insurance_providers ]
```

Comment:
Interesting tidbit, Ruby has another way of building an array of symbols:

    sections = %i(allergies care_goals conditions encounters immunizations medical_equipment medications procedures insurance_providers)

It's slightly shorter but also slightly more opaque to people who don't know Ruby well, so I don't have a real preference, just thought I'd let you know about it.